### DISCUSSION TITLE
It allows sending simultaneous request when using a cache store

### DISCUSSION THREAD

**Discussion by bruno-rodrigues** [spatie/guzzle-rate-limiter-middleware#11] [Ideas]
Hey all, thanks for everything you guys do.

Just wanted to let you know that I tried setting up a cache store and dispatching external requests from different terminal windows "at the same time", using 1 request per minute.

It works after the first request is sent but then all other pending requests are dispatched at the same time. 
```
[2020-09-01 12:18:37] local.DEBUG:
REQUEST: GET [redacted] HTTP/1.1
--
[2020-09-01 12:19:37] local.DEBUG:
REQUEST: GET [redacted] HTTP/1.1
--
[2020-09-01 12:19:37] local.DEBUG:
REQUEST: [redacted] HTTP/1.1
--
[2020-09-01 12:19:37] local.DEBUG:
REQUEST: GET [redacted] HTTP/1.1
```

**Comment by dmyers** [comment]
I think you need to use an atomic locking mechanism so that if you were to have PHP processes that ran at the same time somehow it would be only allowable to the limit you set.

Laravel cache example:
https://laravel.com/docs/7.x/cache#atomic-locks

**Comment by sebastiandedeyne** [comment]
I agree that this isn't ideal. I'm not actively working on this package anymore, but I'd accept PR's that bring improvements here.

**Reply by kdefives** [reply]
I tried to found a solution by adding a concept of "lock mecanism" using another Store (FS adapter), in order to ensure that public function handle(callable $callback) from RateLimiter.php class is executed once a time. By this way we could ensure that handle function is reading in the Store using delayUntilNextRequest never in the same time if we have multiple process in parallel. But i do not find solution currently to implement that.

Because the problem is:
process 1 is executing delayUntilNextRequest() and process 2 is executing delayUntilNextRequest() in same time, then the sleep will be almost the same.
So, after the sleep, this two process are executing the HTTP request even if the rate limit is set to 1.
Maybe, before the sleep, we should push in the Store the information that some processes are pending to be processed be.

For the moment i do not have solution. :/

**Comment by kdefives** [comment]
Same issue with Symfony 5.2 using FileSystem cache : https://github.com/spatie/guzzle-rate-limiter-middleware/discussions/21

As you can see below a test with a rate limit set to 1 per sec. Both programs are waiting correclty at the begining. But 1 minute after the first call, both program are executing http call in same time... Normally only one one two should be executed as we set the limit to 1.

```cli
root@7b625315923e:/var/www# php bin/console app:sync-stats -v
09:15:09 NOTICE    [app] Command app:sync-stats started...
09:15:09 NOTICE    [app] Database variable wait_time set to 3600 for this session
09:15:09 NOTICE    [app] API call to get val content...
09:15:10 NOTICE    [app] API call executed.
09:15:10 NOTICE    [app] API call to get active shard region...
09:16:10 NOTICE    [app] API call executed.
09:16:10 NOTICE    [app] API call to get match list...
```

```cli
root@7b625315923e:/var/www# php bin/console app:newcomer-sync-stats -v
09:15:30 NOTICE    [app] Command app:newcomer-sync-stats started...
09:15:30 NOTICE    [app] Database variable wait_time set to 3600 for this session
09:15:30 NOTICE    [app] API call to get val content...
09:16:10 NOTICE    [app] API call executed.
09:16:10 NOTICE    [app] API call to get active shard region...
```
